File: internal/models/permission.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type Permission struct {
	ID          uuid.UUID
	Name        string
	Description string
	Action      PermissionAction
	RoleLevel   Role
	CreatedAt   int64
	UpdatedAt   int64
	DeletedAt   *int64
}

func (p *Permission) Response() *responses.PermissionResponse {
	return &responses.PermissionResponse{
		ID:          p.ID.String(),
		Name:        p.Name,
		Description: p.Description,
		Action:      p.Action.String(),
		RoleLevel:   p.RoleLevel.String(),
		CreatedAt:   p.CreatedAt,
		UpdatedAt:   p.UpdatedAt,
	}
}

func (p *Permission) DBUpdate() *storage.UpdatePermissionParams {
	return &storage.UpdatePermissionParams{
		ID:          p.ID,
		Name:        p.Name,
		Description: p.Description,
	}
}

func PermissionFromDB(row *storage.DivaPermission) *Permission {
	return &Permission{
		ID:          row.ID,
		Name:        row.Name,
		Description: row.Description,
		Action:      PermissionActionFromString(row.Action),
		RoleLevel:   RoleFromDB(row.RoleLevel),
		CreatedAt:   row.CreatedAt,
		UpdatedAt:   row.UpdatedAt,
		DeletedAt:   row.DeletedAt,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/permission/Permission.kt
@file:OptIn(ExperimentalJsExport::class)
@file:DivaJsExport

package io.github.juevigrace.diva.lib.models.permission

import io.github.juevigrace.diva.core.DivaJsExport
import io.github.juevigrace.diva.core.None
import io.github.juevigrace.diva.core.Option
import io.github.juevigrace.diva.lib.models.api.permission.PermissionResponse
import io.github.juevigrace.diva.lib.models.user.Role
import io.github.juevigrace.diva.lib.models.user.safeRole
import kotlin.js.ExperimentalJsExport
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class Permission(
    val id: Uuid,
    val name: String,
    val description: String,
    val action: PermissionAction,
    val roleLevel: Role,
    val createdAt: Instant,
    val updatedAt: Instant,
    val deletedAt: Option<Instant> = None,
) {
    companion object {
        fun fromResponse(response: PermissionResponse): Permission {
            return Permission(
                id = Uuid.parse(response.id),
                name = response.name,
                description = response.description,
                action = safePermissionAction(response.action),
                roleLevel = safeRole(response.roleLevel),
                createdAt = Instant.fromEpochMilliseconds(response.createdAt),
                updatedAt = Instant.fromEpochMilliseconds(response.updatedAt),
            )
        }
    }
}
